### PR TITLE
gh-90005: Port _ctypes to PY_STDLIB_MOD

### PR DESCRIPTION
**Pull request by tiran**
Co-authored-by: Erlend Egeberg Aasland <erlend.aasland@innova.no>

https://bugs.python.org/issue45847

* Issue: gh-90005

Automerge-Triggered-By: GH:tiran

### CODE REVIEW

**Comment by arhadthedev** [comment]
Is is a harbinger of finally devendored libffi? Currently, CPython ships 1.20 ([1998](http://web.archive.org/web/20020128223345/http://sourceware.org/libffi/)) while Debian LTS [supplies](https://packages.debian.org/stretch/libffi-dev) 3.2.1 ([2014](https://github.com/libffi/libffi/commit/20562ac0427c3578250d04c6e34fb0127d4551cf)).

**Comment by tiran** [comment]
ctypes tests are failing on ARM64 macOS

**Comment by ronaldoussoren** [comment]
I'll try to look at the PR over the weekend.

As a general statement I'd prefer to use the system install of libffi on macOS because Apple includes some magic in its version to ensure libffi closures work properly with their code signature system (in particular some magic to ensure that you don't need the "JIT" entitlement to use closures).

**Comment by erlend-aasland** [comment]
@ronaldoussoren:
> As a general statement I'd prefer to use the system install of libffi on macOS [...]

So we can let go of Modules/_ctypes/libffi_osx? From the README files in the Mac directory, it seems like we're _not_ using the system provided libffi when building the official installers?

**Comment by erlend-aasland** [comment]
Perhaps @kulikjak could help review the SunOS parts of this PR.

**Comment by erlend-aasland** [comment]
Changing the title prefix from `bpo-45847` to `gh-90005` should make bedevere happy.

**Comment by erlend-aasland** [comment]
Regarding the `--with-system-ffi` switch; AFAICT, it now behaves like this:

- on non-Darwin systems, it is ignored; `--with-system-ffi=yes` is set implicitly
- on Darwin:
  + `--with-system-ffi=yes` now checks pkg-config and standard search paths
  + `--with-system-ffi=no` checks SDK root

AFAICS, it is not possible to manually override C preprocessor and linker flags using `LIBFFI_CFLAGS` and `LIBFFI_LIBS` if pkg-config fail.

**Comment by kulikjak** [comment]
> Perhaps @kulikjak could help review the SunOS parts of this PR.

Sure! The PR looks good, and I tried compiling it on Solaris SPARC, and running the Python test suite and found no issues.

**Comment by ronaldoussoren** [comment]
> I have no clue how the SDK stuff works on macOS. In order to land the PR I would need some assistance from @ned-deily or @ronaldoussoren . Or can we drop the SDK and use system-ffi on all system?

AFAIK the embedded copy of libffi for macOS is only used for versions of macOS before 10.9. I no longer care at all for that ancient versions, the only reason I care a little about 10.9 support is that makes the Python installer maximally usable.    We started shipping a local fork of libffi (based on work in PyObjC) because upstream didn't support macOS properly due to some differences in the macOS ABI and the SYSV ABI (IIRC related to alignment of arguments). That's no longer relevant.

Personally I'd say that we should drop the local copy of libffi, and just use system-ffi all systems. 


On macOS I have a strong preference for using the system install of libffi because it contains some tricks to play nice with codesigning entitlements, but in the end that's a choice that should be made by whoever builds a binary.  That does mean that we cannot rely on just pkg-config to find the system-ffi as macOS doesn't ship with that tool.

**Comment by ronaldoussoren** [comment]
```
checking for --with-system-ffi... no
checking ffi.h usability... yes
checking ffi.h presence... no
configure: WARNING: ffi.h: accepted by the compiler, rejected by the preprocessor!
configure: WARNING: ffi.h: proceeding with the compiler's result
```

Without specifying the --with-system-ffi I get the warning above when building on a macOS 12.3 system, and get a similar warning about sqlite3. Both extensions build correctly, but this warning could confuse users.

I haven't looked in detail at the PR yet, nor have I tested on a system that does have pkg-config.   Building pkg-config without using a system like homebrew is ... interesting.

**Comment by ronaldoussoren** [comment]
> I will look at and test the PR soon. I also am looking at options for pkg-config.

There is a (partial) python implementation of pkg-config on on PyPI that might be helpful (ignoring the bootstrapping issues when using it to build Python on a recent macOS machine).  On the other hand, just building pkg-config itself shouldn't be too hard, I just haven't gotten around to writing the scripting to install it and its dependencies on my machine yet.

**Comment by tiran** [comment]
There is also https://github.com/pkgconf/pkgconf

**Comment by ned-deily** [comment]
> There is also https://github.com/pkgconf/pkgconf

Yes, that is what I have been playing with. It has a number of advantages over the standard pkg-config including no build dependencies and less restrictive licensing.

**Comment by tiran** [comment]
Users without pkg-config can fall back to ``LIBFFI_CFLAGS`` and ``LIBFFI_LIBS`` env vars, too.

**Comment by tiran** [comment]
@ned-deily @ronaldoussoren I would like to merge this PR rather sooner than later into main. It is required to test ctypes on wasm32-emscripten and will allow us to remove setup.py. Tests are passing on macOS.

File: configure.ac [hunk 1]
```diff
@@ -3564,12 +3599,60 @@ else
...
+    ],
+    [sunos5], [AS_VAR_APPEND([LIBFFI_LIBS], [" -mimpure-text"])]
+  )
+  AS_VAR_IF([ctypes_malloc_closure], [yes], [
+    MODULE__CTYPES_MALLOC_CLOSURE=_ctypes/malloc_closure.c
+    AS_VAR_APPEND([LIBFFI_CFLAGS], [" -DUSING_MALLOC_CLOSURE_DOT_C=1"])
+  ])
+  AC_SUBST([MODULE__CTYPES_MALLOC_CLOSURE])
+
+  dnl HAVE_LIBDL: for dlopen, see bpo-32647
```
Comment (erlend-aasland) [inline configure.ac]: Nit

```suggestion
  dnl HAVE_LIBDL: for dlopen, see gh-76828
```

**Review by erlend-aasland — APPROVED** [review]

**Pull request merged**